File: packages/domain/src/plate-score.ts
/**
 * Plate Score — a deterministic 0-100 rating of a single meal.
 *
 * Design rules, in order of importance:
 *
 * 1. **Pure arithmetic, never an LLM.** The same meal always scores the same, so
 *    the number is reproducible, free to recompute, and defensible.
 * 2. **Only one axis needs personal data.** Macro balance, protein density and
 *    fiber density describe the food itself; calorie fit needs a daily target.
 *    Users without a health target get the other three renormalised rather than
 *    a score against an invented "average" profile. Nothing is fabricated — the
 *    general tier just answers a narrower question.
 * 3. **Missing means unknown, never zero.** An absent nutrient drops its axis
 *    and the remaining weights renormalise. Scoring a missing fiber value as 0
 *    would make every meal logged before micronutrient recovery look fiber-free.
 * 4. **Read per-item nutrition, never summed totals.** `sumTotals` coerces
 *    absent micronutrients to 0, which would silently defeat rule 3.
 * 5. **Medical conditions never change the number.** They may change the wording
 *    around it. Keeping the score to pure nutrition math is what keeps it out of
 *    medical-claim territory.
 */

import type { MacroTotals } from "./types.js";

export type PlateScoreGoal = "maintain" | "lose_gently" | "gain_gently";
export type PlateScoreMealType = "breakfast" | "lunch" | "snack" | "dinner";

export type PlateScoreTier = "general" | "personal";

export type PlateScoreAxis = "calorie_fit" | "protein" | "macro_balance" | "fiber";

export type PlateScoreBand = "excellent" | "good" | "moderate" | "heavy";

/**
 * Present only for users who have completed the health target form. Bodyweight
 * is deliberately not required: protein is judged by density, so the score never
 * needs it.
 */
export type PlateScoreProfile = {
  dailyCalorieTarget: number;
  goal: PlateScoreGoal;
};

export type PlateScoreInput = {
  /** Per-item nutrition. Never pass summed totals: see rule 4 above. */
  items: MacroTotals[];
  mealType: PlateScoreMealType;
  profile?: PlateScoreProfile;
};

/**
 * Why an axis scored what it did.
 *
 * The number alone is not actionable: a portion bar at 0 tells a user nothing
 * about whether they ate too much or too little, or what would change it. The
 * code travels instead of the copy so the wording stays in the app, where it can
 * be adjusted and translated, while the judgement stays here under test.
 */
export type PlateScoreAxisDetail =
  | "on_track"
  | "portion_large"
  | "portion_small"
  | "protein_low"
  | "carb_heavy"
  | "fat_heavy"
  | "fiber_low";

export type PlateScoreAxisResult = {
  axis: PlateScoreAxis;
  /** 0-100 for this axis alone. */
  score: number;
  /** Share of the final score this axis carried, after renormalisation. */
  weight: number;
  detail: PlateScoreAxisDetail;
};

export type PlateScoreResult = {
  score: number;
  band: PlateScoreBand;
  tier: PlateScoreTier;
  axes: PlateScoreAxisResult[];
  /** Axes skipped because the data they need was unavailable. */
  skipped: PlateScoreAxis[];
};

/**
 * Every tunable number in one place.
 *
 * The *shape* of the algorithm is fixed — sum items, score four axes,
 * renormalise weights, weighted average, map to a band — but the numbers are
 * conventions rather than facts. Keeping them in a policy object means the API
 * can serve them from runtime config and the mobile app can compute an
 * identical score locally, so the review screen updates instantly while the
 * values stay tunable from the backend without an app release.
 */
export type PlateScorePolicy = {
  /** Base weights. Inactive axes drop out and the rest renormalise to 100. */
  weights: Record<PlateScoreAxis, number>;
  /** Expected share of the day's calories, used only by the calorie-fit axis. */
  mealShare: Record<PlateScoreMealType, number>;
  /** How far above the expected share still scores above zero. */
  calorieTolerance: Record<PlateScoreMealType, number>;
  /** Protein g per 1000 kcal considered adequate, by goal. */
  proteinDensityTarget: Record<PlateScoreGoal, number>;
  /** Protein target when no goal is known. */
  generalProteinDensityTarget: number;
  /** Fiber g per 1000 kcal; the widely used adequacy guideline is 14. */
  fiberDensityTarget: number;
  /** Acceptable share-of-energy bands, and how hard to punish straying. */
  macroBands: {
    proteinPct: { min: number; max: number };
    carbsPct: { min: number; max: number };
    fatPct: { min: number; max: number };
    penaltyMultiplier: number;
  };
  /** Lower bound of each band. Anything below `moderate` is "heavy". */
  bandCutoffs: { excellent: number; good: number; moderate: number };
};

export const defaultPlateScorePolicy: PlateScorePolicy = {
  weights: { calorie_fit: 25, protein: 30, macro_balance: 25, fiber: 20 },
  mealShare: { breakfast: 0.25, lunch: 0.35, dinner: 0.3, snack: 0.1 },
  // Snacks are deliberately loose: a 300 kcal snack is normal, and a tight band
  // would penalise ordinary eating.
  calorieTolerance: { breakfast: 0.35, lunch: 0.35, dinner: 0.35, snack: 0.7 },
  // Density rather than absolute grams. Judging a meal against a share of a
  // daily gram target punishes small meals: a 95 kcal chai and biscuit would be
  // asked for 20 g of protein. Cutting calories raises the target because
  // protein matters more when energy is scarce.
  proteinDensityTarget: { maintain: 40, lose_gently: 50, gain_gently: 45 },
  generalProteinDensityTarget: 40,
  fiberDensityTarget: 14,
  macroBands: {
    proteinPct: { min: 15, max: 35 },
    carbsPct: { min: 40, max: 65 },
    fatPct: { min: 20, max: 35 },
    penaltyMultiplier: 1.5,
  },
  bandCutoffs: { excellent: 85, good: 70, moderate: 50 },
};

const clamp = (value: number, min = 0, max = 100): number => Math.max(min, Math.min(max, value));

const round = (value: number, decimals = 1): number => {
  const factor = 10 ** decimals;
  return Math.round((value + Number.EPSILON) * factor) / factor;
};

const isUsable = (value: number | undefined): value is number =>
  value !== undefined && Number.isFinite(value) && value >= 0;

/**
 * Sums macros across items while preserving "unknown".
 *
 * A micronutrient counts only when at least one item reported it; if none did,
 * the total stays undefined so the matching axis is skipped rather than scored
 * as zero. Items that omit it are treated as contributing nothing rather than
 * invalidating the whole meal, which is the pragmatic reading when the AI
 * returns fiber for the dal but not the pickle.
 */
const sumForScoring = (items: MacroTotals[]) => {
  let calories = 0;
  let proteinG = 0;
  let carbsG = 0;
  let fatG = 0;
  let fiberG: number | undefined;

  for (const item of items) {
    calories += isUsable(item.calories) ? item.calories : 0;
    proteinG += isUsable(item.proteinG) ? item.proteinG : 0;
    carbsG += isUsable(item.carbsG) ? item.carbsG : 0;
    fatG += isUsable(item.fatG) ? item.fatG : 0;
    if (isUsable(item.fiberG)) fiberG = (fiberG ?? 0) + item.fiberG;
  }

  return { calories, proteinG, carbsG, fatG, fiberG };
};

/**
 * One-sided calorie fit: full marks at or below the expected share, declining
 * only above it.
 *
 * Eating less than a nominal meal share is not a fault. People eat a light
 * breakfast and a bigger lunch, and a meal score has no visibility into the rest
 * of the day. Penalising a 339 kcal breakfast the way an oversized one is
 * penalised told users their normal eating was "heavy", which was wrong.
 */
const calorieFitScore = (actual: number, expected: number, tolerance: number): number => {
  if (expected <= 0) return 0;
  if (actual <= expected) return 100;
  const excess = (actual - expected) / expected;
  if (excess >= tolerance) return 0;
  return clamp(100 * (1 - excess / tolerance));
};

/**
 * Rises to 100 at the target and stays there. Used for protein and fiber, where
 * exceeding the target is not a fault and should not cost points.
 */
const adequacyScore = (actual: number, target: number): number => {
  if (target <= 0) return 0;
  return clamp((actual / target) * 100);
};

/**
 * Distance from a healthy macro split, expressed as share of energy.
 *
 * Bands are deliberately wide: this rewards a plate that is not dominated by a
 * single macro, rather than enforcing one diet's ideal ratio.
 */
const macroBalanceScore = (
  calories: number,
  proteinG: number,
  carbsG: number,
  fatG: number,
  bands: PlateScorePolicy["macroBands"],
): number => {
  if (calories <= 0) return 0;

  const proteinPct = ((proteinG * 4) / calories) * 100;
  const carbsPct = ((carbsG * 4) / calories) * 100;
  const fatPct = ((fatG * 9) / calories) * 100;

  const penalty = (pct: number, low: number, high: number): number => {
    if (pct < low) return low - pct;
    if (pct > high) return pct - high;
    return 0;
  };

  const total =
    penalty(proteinPct, bands.proteinPct.min, bands.proteinPct.max) +
    penalty(carbsPct, bands.carbsPct.min, bands.carbsPct.max) +
    penalty(fatPct, bands.fatPct.min, bands.fatPct.max);

  return clamp(100 - total * bands.penaltyMultiplier);
};

/** Names the macro furthest outside its band, so the copy can be specific. */
const macroBalanceDetail = (
  calories: number,
  proteinG: number,
  carbsG: number,
  fatG: number,
  bands: PlateScorePolicy["macroBands"],
): PlateScoreAxisDetail => {
  if (calories <= 0) return "on_track";

  const carbsOver = ((carbsG * 4) / calories) * 100 - bands.carbsPct.max;
  const fatOver = ((fatG * 9) / calories) * 100 - bands.fatPct.max;
  const proteinUnder = bands.proteinPct.min - ((proteinG * 4) / calories) * 100;

  const worst = Math.max(carbsOver, fatOver, proteinUnder);
  if (worst <= 0) return "on_track";
  if (worst === fatOver) return "fat_heavy";
  if (worst === carbsOver) return "carb_heavy";
  return "protein_low";
};

const bandFor = (score: number, cutoffs: PlateScorePolicy["bandCutoffs"]): PlateScoreBand => {
  if (score >= cutoffs.excellent) return "excellent";
  if (score >= cutoffs.good) return "good";
  if (score >= cutoffs.moderate) return "moderate";
  return "heavy";
};

/**
 * Scores a meal. Returns undefined only when there is nothing to score at all,
 * so callers can hide the card rather than render a meaningless zero.
 */
export const calculatePlateScore = (
  input: PlateScoreInput,
  policy: PlateScorePolicy = defaultPlateScorePolicy,
): PlateScoreResult | undefined => {
  const totals = sumForScoring(input.items);
  if (totals.calories <= 0) return undefined;

  const active: Array<{ axis: PlateScoreAxis; score: number; detail: PlateScoreAxisDetail }> = [];
  const skipped: PlateScoreAxis[] = [];
  const profile = input.profile;

  // Calorie fit — the only axis that needs personal data.
  if (profile && profile.dailyCalorieTarget > 0) {
    const expected = profile.dailyCalorieTarget * policy.mealShare[input.mealType];
    const score = calorieFitScore(
      totals.calories,
      expected,
      policy.calorieTolerance[input.mealType],
    );
    // "Small" is never a penalty, so it is reported only to explain the bar,
    // never to suggest the user should have eaten more.
    const detail: PlateScoreAxisDetail =
      totals.calories > expected
        ? "portion_large"
        : totals.calories < expected * 0.6
          ? "portion_small"
          : "on_track";
    active.push({ axis: "calorie_fit", score, detail });
  } else {
    skipped.push("calorie_fit");
  }

  // Protein — density in both tiers, so portion size never distorts it. The
  // goal shifts the target rather than the measure.
  const proteinTarget = profile
    ? policy.proteinDensityTarget[profile.goal]
    : policy.generalProteinDensityTarget;
  const proteinPer1000 = (totals.proteinG / totals.calories) * 1000;
  const proteinScore = adequacyScore(proteinPer1000, proteinTarget);
  active.push({
    axis: "protein",
    score: proteinScore,
    detail: proteinScore >= 80 ? "on_track" : "protein_low",
  });

  active.push({
    axis: "macro_balance",
    score: macroBalanceScore(
      totals.calories,
      totals.proteinG,
      totals.carbsG,
      totals.fatG,
      policy.macroBands,
    ),
    detail: macroBalanceDetail(
      totals.calories,
      totals.proteinG,
      totals.carbsG,
      totals.fatG,
      policy.macroBands,
    ),
  });

  // Fiber — skipped entirely when unknown, never scored as zero.
  if (totals.fiberG !== undefined) {
    const per1000 = (totals.fiberG / totals.calories) * 1000;
    const fiberScore = adequacyScore(per1000, policy.fiberDensityTarget);
    active.push({
      axis: "fiber",
      score: fiberScore,
      detail: fiberScore >= 80 ? "on_track" : "fiber_low",
    });
  } else {
    skipped.push("fiber");
  }

  const weightSum = active.reduce((sum, entry) => sum + policy.weights[entry.axis], 0);
  if (weightSum <= 0) return undefined;

  const axes: PlateScoreAxisResult[] = active.map((entry) => ({
    axis: entry.axis,
    score: round(entry.score),
    weight: round((policy.weights[entry.axis] / weightSum) * 100),
    detail: entry.detail,
  }));

  const score = Math.round(
    active.reduce((sum, entry) => sum + entry.score * policy.weights[entry.axis], 0) / weightSum,
  );

  return {
    score,
    band: bandFor(score, policy.bandCutoffs),
    tier: profile ? "personal" : "general",
    axes,
    skipped,
  };
};

File: apps/mobile/lib/src/models/plate_score.dart
import 'meal.dart';

/// Plate Score — a deterministic 0-100 rating of a single meal.
///
/// This is a deliberate port of `packages/domain/src/plate-score.ts`. The two
/// implementations are pinned to a shared fixture
/// (`packages/domain/fixtures/plate-score-vectors.json`) that both test suites
/// assert against, so they cannot drift apart silently.
///
/// Why a local copy exists at all: the review screen is an editing screen, and
/// recomputing on the server would mean a network round-trip per portion tweak.
/// The *shape* of the algorithm lives here; every tunable number arrives from
/// the API in `bootstrap.plateScorePolicy`, so the weights and targets stay
/// adjustable from the backend without an app release.
///
/// The API remains authoritative for saved meals. This is used for the live
/// preview while a meal is still being edited.
enum PlateScoreAxis { calorieFit, protein, macroBalance, fiber }

enum PlateScoreBand { excellent, good, moderate, heavy }

enum PlateScoreTier { general, personal }

enum PlateScoreGoal { maintain, loseGently, gainGently }

extension PlateScoreAxisWire on PlateScoreAxis {
  String get wireName => switch (this) {
    PlateScoreAxis.calorieFit => 'calorie_fit',
    PlateScoreAxis.protein => 'protein',
    PlateScoreAxis.macroBalance => 'macro_balance',
    PlateScoreAxis.fiber => 'fiber',
  };

  static PlateScoreAxis? fromWire(String value) => switch (value) {
    'calorie_fit' => PlateScoreAxis.calorieFit,
    'protein' => PlateScoreAxis.protein,
    'macro_balance' => PlateScoreAxis.macroBalance,
    'fiber' => PlateScoreAxis.fiber,
    _ => null,
  };
}

extension PlateScoreBandWire on PlateScoreBand {
  String get wireName => name;

  static PlateScoreBand fromWire(String value) => switch (value) {
    'excellent' => PlateScoreBand.excellent,
    'good' => PlateScoreBand.good,
    'moderate' => PlateScoreBand.moderate,
    _ => PlateScoreBand.heavy,
  };
}

extension PlateScoreTierWire on PlateScoreTier {
  String get wireName => name;

  static PlateScoreTier fromWire(String value) =>
      value == 'personal' ? PlateScoreTier.personal : PlateScoreTier.general;
}

extension PlateScoreGoalWire on PlateScoreGoal {
  String get wireName => switch (this) {
    PlateScoreGoal.maintain => 'maintain',
    PlateScoreGoal.loseGently => 'lose_gently',
    PlateScoreGoal.gainGently => 'gain_gently',
  };

  static PlateScoreGoal fromWire(String? value) => switch (value) {
    'lose_gently' => PlateScoreGoal.loseGently,
    'gain_gently' => PlateScoreGoal.gainGently,
    _ => PlateScoreGoal.maintain,
  };
}

class PlateScoreRange {
  const PlateScoreRange({required this.min, required this.max});

  final double min;
  final double max;

  factory PlateScoreRange.fromJson(
    Map<String, dynamic>? json,
    PlateScoreRange fallback,
  ) {
    if (json == null) return fallback;
    return PlateScoreRange(
      min: (json['min'] as num?)?.toDouble() ?? fallback.min,
      max: (json['max'] as num?)?.toDouble() ?? fallback.max,
    );
  }
}

/// Every tunable number, mirroring `PlateScorePolicy` in the domain package.
///
/// Defaults match the shipped TypeScript values so an older API that does not
/// send a policy still scores identically.
class PlateScorePolicy {
  const PlateScorePolicy({
    this.weights = const {
      PlateScoreAxis.calorieFit: 25,
      PlateScoreAxis.protein: 30,
      PlateScoreAxis.macroBalance: 25,
      PlateScoreAxis.fiber: 20,
    },
    this.mealShare = const {
      MealType.breakfast: 0.25,
      MealType.lunch: 0.35,
      MealType.dinner: 0.3,
      MealType.snack: 0.1,
    },
    this.calorieTolerance = const {
      MealType.breakfast: 0.35,
      MealType.lunch: 0.35,
      MealType.dinner: 0.35,
      MealType.snack: 0.7,
    },
    this.proteinDensityTarget = const {
      PlateScoreGoal.maintain: 40,
      PlateScoreGoal.loseGently: 50,
      PlateScoreGoal.gainGently: 45,
    },
    this.generalProteinDensityTarget = 40,
    this.fiberDensityTarget = 14,
    this.proteinPct = const PlateScoreRange(min: 15, max: 35),
    this.carbsPct = const PlateScoreRange(min: 40, max: 65),
    this.fatPct = const PlateScoreRange(min: 20, max: 35),
    this.penaltyMultiplier = 1.5,
    this.excellentCutoff = 85,
    this.goodCutoff = 70,
    this.moderateCutoff = 50,
  });

  final Map<PlateScoreAxis, double> weights;
  final Map<MealType, double> mealShare;
  final Map<MealType, double> calorieTolerance;
  final Map<PlateScoreGoal, double> proteinDensityTarget;
  final double generalProteinDensityTarget;
  final double fiberDensityTarget;
  final PlateScoreRange proteinPct;
  final PlateScoreRange carbsPct;
  final PlateScoreRange fatPct;
  final double penaltyMultiplier;
  final double excellentCutoff;
  final double goodCutoff;
  final double moderateCutoff;

  static const fallback = PlateScorePolicy();

  /// Parses the policy served in bootstrap. Any missing field falls back to the
  /// bundled default rather than to zero, so a partial payload never produces a
  /// nonsense score.
  factory PlateScorePolicy.fromJson(Map<String, dynamic>? json) {
    if (json == null) return fallback;
    const base = fallback;

    double pick(Map<String, dynamic>? source, String key, double fallbackValue) =>
        (source?[key] as num?)?.toDouble() ?? fallbackValue;

    final weights = json['weights'] as Map<String, dynamic>?;
    final shares = json['mealShare'] as Map<String, dynamic>?;
    final tolerance = json['calorieTolerance'] as Map<String, dynamic>?;
    final protein = json['proteinDensityTarget'] as Map<String, dynamic>?;
    final bands = json['macroBands'] as Map<String, dynamic>?;
    final cutoffs = json['bandCutoffs'] as Map<String, dynamic>?;

    return PlateScorePolicy(
      weights: {
        for (final axis in PlateScoreAxis.values)
          axis: pick(weights, axis.wireName, base.weights[axis]!),
      },
      mealShare: {
        for (final type in MealType.values)
          type: pick(shares, type.name, base.mealShare[type]!),
      },
      calorieTolerance: {
        for (final type in MealType.values)
          type: pick(tolerance, type.name, base.calorieTolerance[type]!),
      },
      proteinDensityTarget: {
        for (final goal in PlateScoreGoal.values)
          goal: pick(protein, goal.wireName, base.proteinDensityTarget[goal]!),
      },
      generalProteinDensityTarget: pick(
        json,
        'generalProteinDensityTarget',
        base.generalProteinDensityTarget,
      ),
      fiberDensityTarget: pick(json, 'fiberDensityTarget', base.fiberDensityTarget),
      proteinPct: PlateScoreRange.fromJson(
        bands?['proteinPct'] as Map<String, dynamic>?,
        base.proteinPct,
      ),
      carbsPct: PlateScoreRange.fromJson(
        bands?['carbsPct'] as Map<String, dynamic>?,
        base.carbsPct,
      ),
      fatPct: PlateScoreRange.fromJson(
        bands?['fatPct'] as Map<String, dynamic>?,
        base.fatPct,
      ),
      penaltyMultiplier: pick(bands, 'penaltyMultiplier', base.penaltyMultiplier),
      excellentCutoff: pick(cutoffs, 'excellent', base.excellentCutoff),
      goodCutoff: pick(cutoffs, 'good', base.goodCutoff),
      moderateCutoff: pick(cutoffs, 'moderate', base.moderateCutoff),
    );
  }
}

class PlateScoreAxisResult {
  const PlateScoreAxisResult({
    required this.axis,
    required this.score,
    required this.weight,
  });

  final PlateScoreAxis axis;
  final double score;
  final double weight;
}

class PlateScoreProfile {
  const PlateScoreProfile({required this.dailyCalorieTarget, required this.goal});

  final int dailyCalorieTarget;
  final PlateScoreGoal goal;
}

class PlateScore {
  const PlateScore({
    required this.score,
    required this.band,
    required this.tier,
    required this.axes,
    required this.skipped,
  });

  final int score;
  final PlateScoreBand band;
  final PlateScoreTier tier;
  final List<PlateScoreAxisResult> axes;
  final List<PlateScoreAxis> skipped;

  /// Reads a score computed by the API. Returns null when absent, so an older
  /// API response simply renders no score rather than a fabricated zero.
  static PlateScore? fromJson(Map<String, dynamic>? json) {
    if (json == null) return null;
    final score = (json['score'] as num?)?.round();
    if (score == null) return null;

    return PlateScore(
      score: score,
      band: PlateScoreBandWire.fromWire(json['band'] as String? ?? 'heavy'),
      tier: PlateScoreTierWire.fromWire(json['tier'] as String? ?? 'general'),
      axes: ((json['axes'] as List<dynamic>?) ?? const [])
          .whereType<Map<String, dynamic>>()
          .map((entry) {
            final axis = PlateScoreAxisWire.fromWire(entry['axis'] as String? ?? '');
            if (axis == null) return null;
            return PlateScoreAxisResult(
              axis: axis,
              score: (entry['score'] as num?)?.toDouble() ?? 0,
              weight: (entry['weight'] as num?)?.toDouble() ?? 0,
            );
          })
          .whereType<PlateScoreAxisResult>()
          .toList(),
      skipped: ((json['skipped'] as List<dynamic>?) ?? const [])
          .whereType<String>()
          .map(PlateScoreAxisWire.fromWire)
          .whereType<PlateScoreAxis>()
          .toList(),
    );
  }

  /// True when the user could get a more tailored score by filling in their
  /// health target. Drives the "personalise this" prompt.
  bool get canPersonalise => skipped.contains(PlateScoreAxis.calorieFit);
}

double _clamp(double value) => value.clamp(0, 100).toDouble();

double _round1(double value) => (value * 10).round() / 10;

/// Sums macros while preserving "unknown": a micronutrient counts only when at
/// least one item reported it. Absent stays absent so its axis is skipped rather
/// than scored as zero.
({double calories, double proteinG, double carbsG, double fatG, double? fiberG})
_sumForScoring(List<MacroTotals> items) {
  var calories = 0.0;
  var proteinG = 0.0;
  var carbsG = 0.0;
  var fatG = 0.0;
  double? fiberG;

  for (final item in items) {
    calories += item.calories.toDouble();
    proteinG += item.proteinG;
    carbsG += item.carbsG;
    fatG += item.fatG;
    final fiber = item.fiberG;
    if (fiber != null && fiber >= 0) fiberG = (fiberG ?? 0) + fiber;
  }

  return (
    calories: calories,
    proteinG: proteinG,
    carbsG: carbsG,
    fatG: fatG,
    fiberG: fiberG,
  );
}

/// One-sided: full marks at or below the expected share, declining only above
/// it. Eating less than a nominal meal share is not a fault.
double _calorieFitScore(double actual, double expected, double tolerance) {
  if (expected <= 0) return 0;
  if (actual <= expected) return 100;
  final excess = (actual - expected) / expected;
  if (excess >= tolerance) return 0;
  return _clamp(100 * (1 - excess / tolerance));
}

/// Rises to the target and plateaus: exceeding it is not a fault.
double _adequacyScore(double actual, double target) {
  if (target <= 0) return 0;
  return _clamp((actual / target) * 100);
}

double _macroBalanceScore(
  double calories,
  double proteinG,
  double carbsG,
  double fatG,
  PlateScorePolicy policy,
) {
  if (calories <= 0) return 0;

  final proteinPct = ((proteinG * 4) / calories) * 100;
  final carbsPct = ((carbsG * 4) / calories) * 100;
  final fatPct = ((fatG * 9) / calories) * 100;

  double penalty(double pct, PlateScoreRange range) {
    if (pct < range.min) return range.min - pct;
    if (pct > range.max) return pct - range.max;
    return 0;
  }

  final total =
      penalty(proteinPct, policy.proteinPct) +
      penalty(carbsPct, policy.carbsPct) +
      penalty(fatPct, policy.fatPct);

  return _clamp(100 - total * policy.penaltyMultiplier);
}

PlateScoreBand _bandFor(int score, PlateScorePolicy policy) {
  if (score >= policy.excellentCutoff) return PlateScoreBand.excellent;
  if (score >= policy.goodCutoff) return PlateScoreBand.good;
  if (score >= policy.moderateCutoff) return PlateScoreBand.moderate;
  return PlateScoreBand.heavy;
}

/// Computes a Plate Score locally, for the live preview while a meal is being
/// edited. Returns null when there is nothing to score, so the caller hides the
/// card rather than showing a meaningless zero.
PlateScore? calculatePlateScore({
  required List<MacroTotals> items,
  required MealType mealType,
  PlateScoreProfile? profile,
  PlateScorePolicy policy = PlateScorePolicy.fallback,
}) {
  final totals = _sumForScoring(items);
  if (totals.calories <= 0) return null;

  final active = <(PlateScoreAxis, double)>[];
  final skipped = <PlateScoreAxis>[];

  if (profile != null && profile.dailyCalorieTarget > 0) {
    final expected = profile.dailyCalorieTarget * policy.mealShare[mealType]!;
    active.add((
      PlateScoreAxis.calorieFit,
      _calorieFitScore(totals.calories, expected, policy.calorieTolerance[mealType]!),
    ));
  } else {
    skipped.add(PlateScoreAxis.calorieFit);
  }

  final proteinTarget = profile != null
      ? policy.proteinDensityTarget[profile.goal]!
      : policy.generalProteinDensityTarget;
  final proteinPer1000 = (totals.proteinG / totals.calories) * 1000;
  active.add((PlateScoreAxis.protein, _adequacyScore(proteinPer1000, proteinTarget)));

  active.add((
    PlateScoreAxis.macroBalance,
    _macroBalanceScore(
      totals.calories,
      totals.proteinG,
      totals.carbsG,
      totals.fatG,
      policy,
    ),
  ));

  final fiber = totals.fiberG;
  if (fiber != null) {
    final per1000 = (fiber / totals.calories) * 1000;
    active.add((
      PlateScoreAxis.fiber,
      _adequacyScore(per1000, policy.fiberDensityTarget),
    ));
  } else {
    skipped.add(PlateScoreAxis.fiber);
  }

  final weightSum = active.fold<double>(
    0,
    (sum, entry) => sum + policy.weights[entry.$1]!,
  );
  if (weightSum <= 0) return null;

  final score =
      (active.fold<double>(
                0,
                (sum, entry) => sum + entry.$2 * policy.weights[entry.$1]!,
              ) /
              weightSum)
          .round();

  return PlateScore(
    score: score,
    band: _bandFor(score, policy),
    tier: profile != null ? PlateScoreTier.personal : PlateScoreTier.general,
    axes: active
        .map(
          (entry) => PlateScoreAxisResult(
            axis: entry.$1,
            score: _round1(entry.$2),
            weight: _round1((policy.weights[entry.$1]! / weightSum) * 100),
          ),
        )
        .toList(),
    skipped: skipped,
  );
}
